tests/simple-tests/Tests/Util.hs

```diff
@@ -38,59 +42,112 @@ data MsgMetaData = MsgMetaData
+capnpEncode msgValue meta = do
+  (exitStatus, stdErr, contents) <- runResourceT $ do
+    (hin, hout, herr, phandle) <- interactCapnpWithSchema "encode" (msgSchema meta) [msgType meta]
+    contents <- lift $ do
+      forkIO $ do
+        hPutStr hin msgValue
+        hClose hin
+      hSetBinaryMode hout True
+      BS.hGetContents hout
+    stdErr <- lift $ BS.hGetContents herr
+    exitStatus <- lift $ waitForProcess phandle
+    return (exitStatus, stdErr, contents)
+  case exitStatus of
+    ExitSuccess -> return contents
+    ExitFailure code -> fail ("`capnp encode` failed with exit code " ++ show code ++ ":\n" ++ show stdErr)
 
--- | @capnpDecode msg meta@ runs @capnp encode@ on the message, providing
+-- | @capnpDecode msg meta@ runs @capnp decode@ on the message, providing
 -- the needed metadata and returning the output
 capnpDecode :: BS.ByteString -> MsgMetaData -> IO String
-capnpDecode msgValue meta = runResourceT $ do
-    (hin, hout) <- interactCapnp "decode" meta
-    lift $ do
-        forkIO $ do
-            hSetBinaryMode hin True
-            BS.hPutStr hin msgValue
-            hClose hin
-        ret <- hGetContents hout
-        -- We need to read the whole string in strictly, otherwise hClose may
-        -- happen before we're done; use deepseq to force full evaluation:
-        deepseq ret (return ret)
+capnpDecode encodedMsg meta = do
+  (exitStatus, stdErr, contents) <- runResourceT $ do
+    (hin, hout, herr, phandle) <- interactCapnpWithSchema "decode" (msgSchema meta) [msgType meta]
+    contents <- lift $ do
+      forkIO $ do
+        hSetBinaryMode hin True
+        BS.hPutStr hin encodedMsg
+        hClose hin
+      ret <- hGetContents hout
+      -- We need to read the whole string in strictly, otherwise hClose may
+      -- happen before we're done; use deepseq to force full evaluation:
+      deepseq ret (return ret)
+    stdErr <- lift $ BS.hGetContents herr
+    exitStatus <- lift $ waitForProcess phandle
+    return (exitStatus, stdErr, contents)
+  case exitStatus of
+    ExitSuccess -> return contents
+    ExitFailure code -> fail ("`capnp decode` failed with exit code " ++ show code ++ ":\n" ++ show stdErr)
+
+-- | @capnpCompile msg meta@ runs @capnp compile@ on the schema, providing
+-- the needed metadata and returning the output
+capnpCompile :: String -> String -> IO BS.ByteString
+capnpCompile msgSchema outputArg = do
+  (exitStatus, stdErr, contents) <- runResourceT $ do
+    (hin, hout, herr, phandle) <- interactCapnpWithSchema "compile" msgSchema [outputArg]
+    contents <- lift $ do
+      hSetBinaryMode hout True
+      BS.hGetContents hout
+    stdErr <- lift $ BS.hGetContents herr
+    exitStatus <- lift $ waitForProcess phandle
+    return (exitStatus, stdErr, contents)
+  case exitStatus of
+    ExitSuccess -> return contents
+    ExitFailure code -> fail ("`capnp compile` failed with exit code " ++ show code ++ ":\n" ++ show stdErr)
+
+capnpId :: IO BS.ByteString
```

Comment:
calling out to `capnp id` is excessive -- all it does is generate a random number and print it.